go-kit/metrics/provider/librato/uniquecounter.go

```diff
@@ -0,0 +1,75 @@
+/* Copyright (c) 2018 Salesforce
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+package librato
+
+import (
+	"sync"
+
+	hll "github.com/axiomhq/hyperloglog"
+	xmetrics "github.com/heroku/x/go-kit/metrics"
+)
+
+var (
+	_ xmetrics.UniqueCounter = &HLLCounter{}
+)
+
+// HLLCounter provides a wrapper around a HyperLogLog probabalistic
+// counter, capable of being reported to Librato.
+type HLLCounter struct {
+	Name    string
+	lvs     []string
+	mu      sync.RWMutex
+	counter *hll.Sketch
+}
+
+func NewHLLCounter(name string) *HLLCounter {
+	return &HLLCounter{
+		Name:    name,
+		counter: hll.New(),
+	}
+}
+
+// With returns a new UniqueCounter with the passed in label values merged
+// with the previous label values. The counter value is forked.
+func (c *HLLCounter) With(labelValues ...string) xmetrics.UniqueCounter {
+	nlv := make([]string, len(c.lvs)+len(labelValues), 0)
+	nlv = append(nlv, c.lvs...)
+	nlv = append(nlv, labelValues...)
+
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return &HLLCounter{
+		Name:    c.Name,
+		lvs:     nlv,
+		counter: c.counter.Clone(),
+	}
+}
+
+// Insert counts x as a unique value.
+func (c *HLLCounter) Insert(x []byte) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.counter.Insert(x)
+}
+
+// Estimate provides the counter's current estimate of cardinality.
+func (c *HLLCounter) Estimate() uint64 {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.counter.Estimate()
```

Comment:
That doesn't pan out. My comment on suggesting Estimate was expensive was mostly around keeping the Gauge in sync, e.g., calling `Estimate()` for every `Insert` instead of doing a atomic CAS swap, like the Gauge, or an atomic increment, like the Counter.